Symplectic integration in a user-defined magnetostatic vector potential

cemitch99 opened the pull request:
This PR adds a new element that allows the user to track through a region with a specified magnetostatic vector potential.  Symplectic integration is performed using the exact form of the nonlinear relativistic Hamiltonian.  We use the semiexplicit integrator appearing in:

B. Jayawardana and T. Ohsawa, "Semiexplicit symplectic integrators for non-separable Hamiltonian systems," Math. Comput. 92, pp. 251-281 (2022),
https://doi.org/10.1090/mcom/3778 ([arXiv](https://arxiv.org/pdf/2111.10915))

To do:

- [x] add template for the basic VectorPotential element
- [x] add functions needed for the semiexplicit integrator
- [x] update the map in VectorPotential to use the correct Hamiltonian derivatives
- [x] update user-facing inputs
- [x] add simple benchmark problem(s)
- [x] check the treatment of s-dependence
- [x] add an s-dependent benchmark problem
- [x] add Python bindings
- [x] update documentation
- [x] add support for 4th-order and 6th-order integration

cemitch99 committed:
Add initial element and integrator routines in draft form.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add literals.

cemitch99 committed:
Resolve conflicts, and add comments.

cemitch99 committed:
Merge branch 'development' into add_user_nonlinear_element

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update Integrators.H

cemitch99 committed:
Fix parameter documentation in new integrator step.

cemitch99 committed:
Add both particle copies as arguments to internal push map.

cemitch99 committed:
Clean-up and addition of map for a general Hamiltonian.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add preliminary initialization of VectorPotential element.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Eliminate calls to Print to avoid errors.

cemitch99 committed:
Add maybe_unused to avoid unused variable warnings.

cemitch99 committed:
Attempt to eliminate host-device error.

cemitch99 committed:
Remove debug print statement.

cemitch99 committed:
Update map using Hamiltonian derivatives.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Delete unnecessary comments.

cemitch99 commented on src/elements/VectorPotential.H:
```diff
@@ -0,0 +1,444 @@
+         * @param t temporal coordinate in m
+         * @param zeval lattice position in m
+         * @param[out] tuple containing the vector potential and derivatives
+         */
```
This function is hard-coded at the moment, to represent the test case of a quad.  These quantities we would like to obtain directly from the user using the math parser.

ax3l replied:
Thanks for the annotation! Updated :)

cemitch99 committed:
Comment unused az.

cemitch99 commented:
Note: Some host-device errors appear in CI when running on GPU that will need to be addressed.

cemitch99 committed:
Add normalization of field strength.

ax3l committed:
User-Provided Functions

ax3l committed:
Fix Doxygen

ax3l commented:
> https://github.com/BLAST-ImpactX/impactx/pull/964#issuecomment-3085567714

Fixed GPU compile :) This was just lacking a host device annotation in a function declaration.

ax3l committed:
Start Python Binding

cemitch99 committed:
Clean-up and add quad example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add maybe_unused to avoid unused variable errors.

cemitch99 committed:
Simplify user syntax and add examples to CMake.

cemitch99 committed:
Merge branch 'development' into add_user_nonlinear_element

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

cemitch99 committed:
Update CMakeLists.txt

cemitch99 committed:
Fix treatment of s-dependence.

cemitch99 committed:
Add soft-edge solenoid example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Correct Doxygen in VectorPotential.H

cemitch99 committed:
Rename VectorPotential to MagnetostaticVectorPotential.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Remove unused t-variable from Parser.

cemitch99 committed:
Add draft of element documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add support for 4th-order and 6th-order integration.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_user_nonlinear_element

ax3l committed:
CUDA CI: More Disk Space

Runner gets killed... might use too much disk space or
RAM at runtime...?
```
2025-08-06T16:36:15.7086639Z ##[error]The runner has received a shutdown signal. This can happen when the runner service is stopped, or a manually started runner is canceled.
2025-08-06T16:36:15.7095056Z ##[error]Process completed with exit code 143.
```

ax3l committed:
Small Cleanup

ax3l committed:
CUDA CI: -j 3

Reduce parallelism to avoid job shutdown

ax3l commented:
The tolerances seem to fail with MPI, maybe just need a small relaxation @cemitch99 ?
```
Errors while running CTest
	485 - examples-fodo-vector-potential.py.analysis (Failed)
	491 - examples-exact-quad-vector-potential.py.analysis (Failed)
	512 - examples-solenoid-vector-potential.py.analysis (Failed)
```

ax3l committed:
CUDA CI: -j 2

cemitch99 committed:
Relax tolerance analysis_exact_quad_vector_potential.py

cemitch99 committed:
Relax tolerance analysis_solenoid_vector_potential.py

ax3l commented on src/elements/MagnetostaticVectorPotential.H:
```diff
@@ -0,0 +1,481 @@
+            amrex::ParticleReal dHdpt = 0_prt;
+            amrex::ParticleReal z = zeval;
+
+            // Evaluate the vector potential and its derivatives
+            auto const ax = m_dfunc_ax(x, y, z) * m_scale;
+            auto const ay = m_dfunc_ay(x, y, z) * m_scale;
+            auto const daxdx = m_dfunc_daxdx(x, y, z) * m_scale;
+            auto const daxdy = m_dfunc_daxdy(x, y, z) * m_scale;
+            auto const daydx = m_dfunc_daydx(x, y, z) * m_scale;
+            auto const daydy = m_dfunc_daydy(x, y, z) * m_scale;
+            auto const dazdx = m_dfunc_dazdx(x, y, z) * m_scale;
+            auto const dazdy = m_dfunc_dazdy(x, y, z) * m_scale;
```
@WeiqunZhang I fear we use too many parsers for the CUDA CI (?) :cry:

ax3l replied:
Possible hack to confirm
```suggestion
            // Evaluate the vector potential and its derivatives
            auto const ax = 0_prt;
            auto const ay = 0_prt;
            auto const daxdx = 0_prt;
            auto const daxdy = 0_prt;
            auto const daydx = 0_prt;
            auto const daydy = 0_prt;
            auto const dazdx = 0_prt;
            auto const dazdy = 0_prt;
```

WeiqunZhang replied:
We probably could try to merge them into one Parser. Let me think about it.

WeiqunZhang replied:
Or maybe it does not help. I think we need to force noinline them.

WeiqunZhang replied:
How about this?
```
amrex::GpuArray<amrex::ParserExecutor<3>,8> df{m_dfunc_ax, m_dfunc_ay, ....}; // on host

// on device
amrex::Real results[8];
for (int i = 0; i < 8; ++i) {
    results[i] = df[i](x,y,z) * m_scale;
}
```

WeiqunZhang replied:
You might add a pragma to make sure the loop is not unrolled.

WeiqunZhang replied:
As for the noinline approach, you can add a helper in your code that that marked as noinline. Something like
```
AMREX_GPU_HOST_DEVICE AMREX_NO_INLINE
template <int N, typename... T>
auto call_parser (ParserExecutor<N> const& f, T... xyz)
{
    return f(xyz...);
}
```

ax3l replied:
I think that might be good!

So far, reducing the compile to `-j1` for CUDA helped, but calling a non-inline for this case would be super helpful.

@WeiqunZhang do you like to pus hthis to the PR or a follow-up?

WeiqunZhang replied:
I can put the helper function in amrex. Then we can use it here or in another PR. (I tested it. It did work.)

ax3l replied:
https://github.com/AMReX-Codes/amrex/pull/4606

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -53,4 +64,4 @@ jobs:
           -DAMReX_CUDA_ERROR_CROSS_EXECUTION_SPACE_CALL=ON \
           -DAMReX_CUDA_ERROR_CAPTURE_THIS=ON
-        cmake --build build -j 4
+        cmake --build build -j 2
```
```suggestion
        cmake --build build -j 1
```

ax3l committed:
CUDA CI: -j 1

cemitch99 committed:
Merge branch 'development' into add_user_nonlinear_element

ax3l committed:
CMake: Unify Test Names

ax3l committed:
Relax Tolerances

ax3l committed:
Relax Tolerances

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_user_nonlinear_element

ax3l commented:
Something is still off with the Python tests. Besides the tolerance issues, they seem to run significantly longer than their app/executable counterparts?
```
ctest --test-dir build -R vector
      Start 481: fodo-vector-potential.run
 1/18 Test #481: fodo-vector-potential.run .................   Passed    3.65 sec
      Start 482: fodo-vector-potential.analysis
 2/18 Test #482: fodo-vector-potential.analysis ............   Passed    0.30 sec
      Start 483: fodo-vector-potential.cleanup
 3/18 Test #483: fodo-vector-potential.cleanup .............   Passed    0.01 sec
      Start 484: fodo-vector-potential.py.run
 4/18 Test #484: fodo-vector-potential.py.run ..............   Passed  268.34 sec
      Start 485: fodo-vector-potential.py.analysis
 5/18 Test #485: fodo-vector-potential.py.analysis .........***Failed    0.26 sec
      Start 486: fodo-vector-potential.py.cleanup
 6/18 Test #486: fodo-vector-potential.py.cleanup ..........   Passed    0.01 sec
      Start 487: exact-quad-vector-potential.run
 7/18 Test #487: exact-quad-vector-potential.run ...........   Passed    3.73 sec
      Start 488: exact-quad-vector-potential.analysis
 8/18 Test #488: exact-quad-vector-potential.analysis ......   Passed    0.57 sec
      Start 489: exact-quad-vector-potential.cleanup
 9/18 Test #489: exact-quad-vector-potential.cleanup .......   Passed    0.01 sec
      Start 490: exact-quad-vector-potential.py.run
10/18 Test #490: exact-quad-vector-potential.py.run ........   Passed   11.98 sec
      Start 491: exact-quad-vector-potential.py.analysis
11/18 Test #491: exact-quad-vector-potential.py.analysis ...***Failed    0.57 sec
      Start 492: exact-quad-vector-potential.py.cleanup
12/18 Test #492: exact-quad-vector-potential.py.cleanup ....   Passed    0.01 sec
      Start 508: solenoid-vector-potential.run
13/18 Test #508: solenoid-vector-potential.run .............   Passed    6.24 sec
      Start 509: solenoid-vector-potential.analysis
14/18 Test #509: solenoid-vector-potential.analysis ........   Passed    0.59 sec
      Start 510: solenoid-vector-potential.cleanup
15/18 Test #510: solenoid-vector-potential.cleanup .........   Passed    0.01 sec
      Start 511: solenoid-vector-potential.py.run
16/18 Test #511: solenoid-vector-potential.py.run ..........   Passed   23.76 sec
      Start 512: solenoid-vector-potential.py.analysis
17/18 Test #512: solenoid-vector-potential.py.analysis .....***Failed    0.58 sec
      Start 513: solenoid-vector-potential.py.cleanup
18/18 Test #513: solenoid-vector-potential.py.cleanup ......   Passed    0.01 sec
```

ax3l commented on docs/source/usage/parameters.rst:
```diff
@@ -429,6 +429,39 @@ This element requires these additional parameters:
+This element requires these additional parameters:
+
+* ``<element_name>.ds`` (``float``, in meters) the segment length
+* ``<element_name>.unit`` (``integer``) specification of units for the vector potential (default: ``0``)
```
This is called `unit` while the inputs file (app API) calls it `units`

ax3l replied:
fixed

ax3l committed:
Fix (Some): Py FODO Vector Example

Was not the same input as used in the app example

ax3l committed:
Fix (Some): Py Solenoid Vector Example

Was not the same input as used in the app example

ax3l commented on examples/vector_potential/run_exact_quad_vector_potential.py:
```diff
@@ -0,0 +1,111 @@
+        beta_x=0.002,
+        beta_y=0.002,
+        beta_t=sigma_tau / sigma_p,
+        emitt_x=1.5e-6 / bg,
+        emitt_y=1.5e-6 / bg,
```
This value is numerically `7.665084325966243e-09` but the inputs file reads `7.626114e-9`

ax3l replied:
Clarified in person: the distributed compared against here is at a different point than the input/source.

ax3l committed:
Fix (Some): Py Quad Vector Example

Was not the same input as used in the app example

ax3l commented:
@cemitch99 the three python run files are not yet 100% identical with the app inputs files, which should be the origin of the failing tests. I fixed the things I spotted, but some differences remain.

cemitch99 committed:
Make input_exact_quad_vector_potential.in exactly coincide with run_exact_quad_vector_potential.py.

cemitch99 committed:
Delete irrelevant comments in input_solenoid_vector_potential.in

cemitch99 commented:
The three new tests currently fail only in the case OpenMP / GCC w/ MPI w/ Python.  Note that the execution time (of the Python test) is very long - 32.42 sec for the solenoid example, which is only 0.56 sec on macOS / AppleClang (with similar behavior for the other two tests).  Also, although the solenoid example runs successfully, the initial and final beam moments agree to all digits (and they should not), which appears to indicate that no tracking push was applied to the beam.

ax3l commented:
Thanks! If they run a bit on the longer end, let us add the `slow` label on these (in `CMakeLists.txt`).

cemitch99 committed:
Merge branch 'development' into add_user_nonlinear_element

cemitch99 commented on examples/vector_potential/analysis_fodo_vector_potential.py:
```diff
@@ -0,0 +1,116 @@
+s = rbc["s"]
+sig_x = rbc["sig_x"]
+sig_y = rbc["sig_y"]
+sig_t = rbc["sig_t"]
```
```suggestion
sig_x = rbc["sigma_x"]
sig_y = rbc["sigma_y"]
sig_t = rbc["sigma_t"]
```

cemitch99 committed:
Update examples/vector_potential/analysis_fodo_vector_potential.py

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_user_nonlinear_element

# Conflicts:
#	.github/workflows/cuda.yml
#	examples/CMakeLists.txt

ax3l committed:
Modernize MagnetostaticVectorPotential element

ax3l committed:
Add `finalize()`

ax3l committed:
Fix Warnings

ax3l commented on .github/workflows/cuda.yml:
```diff
@@ -63,4 +63,4 @@ jobs:
           -DAMReX_CUDA_ERROR_CROSS_EXECUTION_SPACE_CALL=ON \
           -DAMReX_CUDA_ERROR_CAPTURE_THIS=ON
-        cmake --build build -j 4
+        cmake --build build -j 1
```
```suggestion
        cmake --build build -j 4
```

ax3l commented on src/initialization/InitElement.cpp:
```diff
@@ -508,6 +508,58 @@ element_name) );
+            int units = 0;
+            int int_order = 2;
+            int mapsteps = mapsteps_default;
+            pp_element.queryAddWithParser("units", units);
```
all other elements call this singular `unit`:
```suggestion
            int unit = 0;
            int int_order = 2;
            int mapsteps = mapsteps_default;
            pp_element.queryAddWithParser("unit", unit);
```
https://impactx.readthedocs.io/en/latest/usage/parameters.html

ax3l commented on src/initialization/InitElement.cpp:
```diff
@@ -508,6 +508,58 @@ element_name) );
+
+            m_lattice.emplace_back(MagnetostaticVectorPotential(
+                ds,
+                units,
```
```suggestion
                unit,
```

ax3l commented on src/python/elements.cpp:
```diff
@@ -2873,6 +2873,95 @@ void init_elements(py::module& m)
+            "Number of integration steps per slice"
+        )
+    ;
+    register_push(py_MagnetostaticVectorPotential);
```
```suggestion
    register_push(py_MagnetostaticVectorPotential);
    register_reverse(py_MagnetostaticVectorPotential);
```

ax3l commented on src/elements/MagnetostaticVectorPotential.H:
```diff
@@ -0,0 +1,454 @@
+         * local frame. Reversing only negates the path-length integration
+         * direction; the user is responsible for the field's z-symmetry.
+         */
+        void reverse () { Thick::reverse(); }
```
To check: complete or is there more to do? (needs a test entry in `tests/python/test_reversibility_elements.py`)

cemitch99 replied:
In principle, the splitting should be symmetric here.  Suggest we add a reversibility test in a follow-up.

ax3l committed:
CUDA CI: Keep `-j 4`

ax3l commented on examples/vector_potential/input_exact_quad_vector_potential.in:
```diff
@@ -0,0 +1,61 @@
+quad2.ds = 0.02903
+quad2.dA_z/dx(x,y,z) = 207.0*x
+quad2.dA_z/dy(x,y,z) = -207.0*y
+quad2.units = 1
```
```suggestion
quad1.unit = 1
quad1.mapsteps = 5

quad2.type = magnetostatic_vector_potential
quad2.ds = 0.02903
quad2.dA_z/dx(x,y,z) = 207.0*x
quad2.dA_z/dy(x,y,z) = -207.0*y
quad2.unit = 1
```

ax3l commented on examples/vector_potential/input_fodo_vector_potential.in:
```diff
@@ -0,0 +1,65 @@
+quad2.ds = 1.0
+quad2.dA_z/dx(x,y,z) = x
+quad2.dA_z/dy(x,y,z) = -y
+quad2.units = 0
```
```suggestion
quad1.unit = 0
quad1.int_order = 4
quad1.mapsteps = 4

drift2.type = drift
drift2.ds = 0.5

quad2.type = magnetostatic_vector_potential
quad2.ds = 1.0
quad2.dA_z/dx(x,y,z) = x
quad2.dA_z/dy(x,y,z) = -y
quad2.unit = 0
```

ax3l commented on examples/vector_potential/input_solenoid_vector_potential.in:
```diff
@@ -0,0 +1,50 @@
+sol1.A_y(x,y,z) = 0.790585*x*((1-2*z)/sqrt(0.111156+(1-2*z)**2)+(1+2*z)/sqrt(0.111156+(1+2*z)**2))
+sol1.dA_x/dy(x,y,z) = -0.790585*((1-2*z)/sqrt(0.111156+(1-2*z)**2)+(1+2*z)/sqrt(0.111156+(1+2*z)**2))
+sol1.dA_y/dx(x,y,z) = 0.790585*((1-2*z)/sqrt(0.111156+(1-2*z)**2)+(1+2*z)/sqrt(0.111156+(1+2*z)**2))
+sol1.units = 1
```
```suggestion
sol1.unit = 1
```

ax3l committed:
units -> unit

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l committed:
Python Bindings Register Reverse

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l commented:
@cemitch99 I updated this PR and test pass now :)

A few smaller inline comments remain.

ax3l commented on src/elements/MagnetostaticVectorPotential.H:
```diff
@@ -0,0 +1,454 @@
+            // in the Hamiltonian.
+            amrex::ParticleReal const pxmech = px - ax;
+            amrex::ParticleReal const pymech = py - ay;
+            amrex::ParticleReal const radical = std::sqrt(
+                1_prt + pt*pt - px*px - py*py - 2_prt*pt*m_ibeta
+            );
```
Codex:
>   - High: src/elements/MagnetostaticVectorPotential.H:348 computes the
    Hamiltonian radical with px*px and py*py, but the derivatives use mechanical
    momenta px - ax and py - ay. For a vector-potential Hamiltonian, the
    denominator should use pxmech/pymech; otherwise elements with nonzero A_x/
    A_y get the wrong nonlinear kinematics.

ax3l commented on src/particles/integrators/Integrators.H:
```diff
@@ -508,6 +508,218 @@ namespace impactx::integrators
+        // Initialize the perturbation vector
+        amrex::SmallVector<amrex::ParticleReal, 6, 1> perturbation{};
+
+        // Iteration loop (implicit step)
+        for (int j = 0; j < max_iterations; j++) {
+
```
Codex:
>   - High: src/particles/integrators/Integrators.H:642 silently leaves particle
    unchanged if the fixed-point/Newton loop never reaches tolerance. This can
    turn a hard integration failure into a no-op push for that particle. I’d
    abort/report non-convergence, and make the tolerance precision-aware.

ax3l replied:
Could be tricky within a GPU kernel. We could make particles invalid with signaling NaNs or whip up a detailed error machinery.

ax3l commented on src/python/elements.cpp:
```diff
@@ -2873,6 +2873,96 @@ void init_elements(py::module& m)
+                 );
+             }
+        )
+        .def("to_dict",
+            [](MagnetostaticVectorPotential const & vp) {
+                return element_dict(
+                    vp,
+                    std::make_pair("unit", vp.m_unit)
+                );
+            }
+        )
```
Codex:
>   - Medium: src/python/elements.cpp:2886 serializes only unit for
    MagnetostaticVectorPotential. ax, ay, derivative formulas, int_order, and
    mapsteps are dropped, so KnownElementsList.to_dicts()/from_dicts()
    reconstructs a different zero-field/default element. The new element should
    also be added to tests/python/test_element_serialization.py.

ax3l commented on docs/source/usage/python.rst:
```diff
@@ -1769,6 +1769,58 @@ Lattice Elements
+
+   .. py:property:: k
+
+      quadrupole strength in 1/m^2 (or T/m)
```
Codex:
>   - Low: docs/source/usage/python.rst:1807 documents a nonexistent k property
    and describes unit as a quad strength unit. Also mapsteps is documented as
    default 5 in docs, while the C++/Python binding default is 10.

cemitch99 replied:
```suggestion
```

ax3l commented on src/elements/MagnetostaticVectorPotential.H:
```diff
@@ -0,0 +1,454 @@
+         * The amrex::Parser instances we own depend on AMReX runtime state, so
+         * we must drop them before AMReX::Finalize() tears that state down.
+         */
+        void
+        finalize ()
+        {
+            DynamicData::registry().erase(m_id);
+        }
```
Codex:
>   - Medium: src/elements/MagnetostaticVectorPotential.H:409 erases the parser
    registry entry during finalize(). Since elements are copied by value into
    the lattice, a user-held Python/C++ element object can still exist with the
    same m_id after sim.finalize(), but its parser data is gone. Reusing that
    element in a later simulation will fail in DynamicData::get(m_id). This also
    contradicts the comment at src/elements/MagnetostaticVectorPotential.H:44
    saying host data persists across simulations. A robust fix would keep the
    original formula strings as durable host data and rebuild AMReX parsers
    after a new initialize cycle.

ax3l replied:
Yes this pattern is wrong, needs to be a `NoFinalize` class. Will update.

ax3l commented on src/particles/integrators/Integrators.H:
```diff
@@ -508,6 +508,218 @@ namespace impactx::integrators
+
+        } else {
+
+           // initialize numerical integration parameters (int_order = 6)
```
Codex:
>   - Medium: src/particles/integrators/Integrators.H:568 treats every int_order
    other than 2 or 4 as sixth order. Because the constructor, input parser, and
    Python setter accept arbitrary integers, int_order=3 or a typo silently
    changes the integration method. This should be else if (int_order == 6) plus
    an error, or validation in the element constructor/parser/setter.

ax3l replied:
I agree, and the setter/constructor should abort on orders that are not 2,4 or 6

ax3l commented on examples/vector_potential/README.rst:
```diff
@@ -0,0 +1,128 @@
+
+   .. literalinclude:: analysis_fodo_vector_potential.py
+      :language: python3
+      :caption: You can copy this file from ``examples/vector_potential/analysis_vector_potential.py``.
```
Codex:
>   - Low: examples/vector_potential/README.rst:45 has a stale caption pointing to
    analysis_vector_potential.py, but the included file is
    analysis_fodo_vector_potential.py.

cemitch99 replied:
```suggestion
      :caption: You can copy this file from ``examples/vector_potential/analysis_fodo_vector_potential.py``.
```

cemitch99 commented on src/elements/MagnetostaticVectorPotential.H:
```diff
@@ -0,0 +1,454 @@
+            amrex::ParticleReal const pxmech = px - ax;
+            amrex::ParticleReal const pymech = py - ay;
+            amrex::ParticleReal const radical = std::sqrt(
+                1_prt + pt*pt - px*px - py*py - 2_prt*pt*m_ibeta
```
```suggestion
                1_prt + pt*pt - pxmech*pxmech - pymech*pymech - 2_prt*pt*m_ibeta
```

cemitch99 replied:
Codex is right.

cemitch99 commented on docs/source/usage/python.rst:
```diff
@@ -1769,6 +1769,58 @@ Lattice Elements
 
+
+.. py:class:: impactx.elements.MagnetostaticVectorPotential(ds, unit=0, ax="0", ay="0", daxdx="0", daxdy="0", daydx="0", daydy="0", dazdx="0", dazdy="0", dx=0, dy=0, rotation=0, aperture_x=0,
+aperture_y=0, int_order=2, mapsteps=5, nslice=1, name=None)
```
```suggestion
aperture_y=0, int_order=2, mapsteps=10, nslice=1, name=None)
```

cemitch99 commented on docs/source/usage/python.rst:
```diff
@@ -1769,6 +1769,58 @@ Lattice Elements
+   :param name: an optional name for the element
+
+   .. py:property:: k
+
```
```suggestion
```

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Chad Mitchell <46825199+cemitch99@users.noreply.github.com>

cemitch99 commented on src/particles/integrators/Integrators.H:
```diff
@@ -508,6 +508,218 @@ namespace impactx::integrators
+           element.map(tau2,particle2,particle1,zeval2,zeval1);
+           element.map(tau1,particle1,particle2,zeval1,zeval2);
+
+        } else {
```
```suggestion
        } else if (int_order==6) {
```

cemitch99 commented on src/particles/integrators/Integrators.H:
```diff
@@ -508,6 +508,218 @@ namespace impactx::integrators
+           element.map(tau3,particle1,particle2,zeval1,zeval2);
+           element.map(tau2,particle2,particle1,zeval2,zeval1);
+           element.map(tau1,particle1,particle2,zeval1,zeval2);
+
```
```suggestion
    } else {
              throw std::runtime_error("Symplectic integrator: order must be 2, 4, or 6.");

```

cemitch99 committed:
Add throw error support to Integrators.H

cemitch99 committed:
Apply suggestion from @cemitch99

cemitch99 committed:
Apply suggestion from @cemitch99

cemitch99 commented on src/particles/integrators/Integrators.H:
```diff
@@ -508,6 +509,220 @@ namespace impactx::integrators
+           element.map(tau2,particle2,particle1,zeval2,zeval1);
+           element.map(tau1,particle1,particle2,zeval1,zeval2);
+    } else {
+              throw std::runtime_error("Symplectic integrator: order must be 2, 4, or 6.");
```
Causes problems on GPU.  We could move the error outside, e.g. at element input.  However, we still need some logic to put here for the else case.

cemitch99 replied:
```suggestion
              // throw std::runtime_error("Symplectic integrator: order must be 2, 4, or 6.");
```

cemitch99 committed:
Comment throw error.
